### PR TITLE
[aborted] Clean up binary element-wise assertions

### PR DESCRIPTION
**Pull request by frreiss**
**Edit: Accidentally messed up my branch. Recreating my changes in a new branch and resubmitting this PR.**

TensorFlow 1.5 added two pieces of useful functionality to the `assert_equals` op. In eager mode, `assert_equals` now prints a more useful error message that pinpoints which elements of the input tensors differ (see [commit 361c55899cb524ca078c65eabdd3d79bfc10c8f9](https://github.com/tensorflow/tensorflow/commit/361c55899cb524ca078c65eabdd3d79bfc10c8f9)). In graph mode, `assert_equals` now evaluates the assertion at graph construction time when both inputs can be evaluated statically (see [commit cfbeafe11d9b86f8685c1c0f97d285885b5a5f1f](https://github.com/tensorflow/tensorflow/commit/cfbeafe11d9b86f8685c1c0f97d285885b5a5f1f)).

This PR ports this additional functionality to the other binary element-wise assertion ops `assert_none_equal`, `assert_less`, `assert_less_equal`, `assert_greater`, and `assert_greater_equal`. 

**Before:**
```
In [1]: import numpy as np 
   ...: import tensorflow as tf 
   ...: tf.enable_eager_execution() 
   ...: zeros = np.zeros(1000) 
   ...: mostly_ones = np.full(1000, 1.) 
   ...: mostly_ones[567] = 0. 
   ...: tf.assert_none_equal(zeros, mostly_ones, summarize=3)            
       
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-1-23dba1a27a31> in <module>
      5 mostly_ones = np.full(1000, 1.)
      6 mostly_ones[567] = 0.
----> 7 tf.assert_none_equal(zeros, mostly_ones, summarize=3)
[...stack trace continues...]

InvalidArgumentError: Expected 'tf.Tensor(False, shape=(), dtype=bool)' to be true. Summarized data: b''
b'Condition x != y did not hold for every single element:'
b'x (shape=(1000,) dtype=float64) = '
0.0, 0.0, 0.0, ...
b'y (shape=(1000,) dtype=float64) = '
1.0, 1.0, 1.0, ...
```
**After:**
```
In [1]: import numpy as np 
   ...: import tensorflow as tf 
   ...: tf.enable_eager_execution() 
   ...: zeros = np.zeros(1000) 
   ...: mostly_ones = np.full(1000, 1.) 
   ...: mostly_ones[567] = 0. 
   ...: tf.assert_none_equal(zeros, mostly_ones, summarize=3)                
                                       
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-1-1c86ae0b9399> in <module>
      5 mostly_ones = np.full(1000, 1.)
      6 mostly_ones[567] = 0.
----> 7 tf.assert_none_equal(zeros, mostly_ones)
[...stack trace continues...]

InvalidArgumentError: Condition x != y did not hold.
Indices of first 1 different values:
[[567]]
Corresponding x values:
[0.]
Corresponding y values:
[0.]
First 3 elements of x:
[0. 0. 0.]
First 3 elements of y:
[1. 1. 1.]
```

The ops `assert_negative`, `assert_non_negative`, `assert_positive`, and `assert_non_positive` also get some of the new functionality, as they are based on `assert_less` and `assert_less_equal`.

**Before:**
```
In [1]: import tensorflow as tf 
   ...: tf.assert_non_negative(-1.)                                             
Out[1]: <tf.Operation 'assert_non_negative/assert_less_equal/Assert/AssertGuard/Merge' type=Merge>
```
**After:**
```
In [1]: import tensorflow as tf 
   ...: tf.assert_non_negative(-1.)                               
                                            
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-1-c2ddbad603d6> in <module>
      1 import tensorflow as tf
----> 2 tf.assert_non_negative(-1.)
[...stack trace continues...]

InvalidArgumentError: 
Condition x >= 0 did not hold element-wise:
x (assert_non_negative/x:0) = 
-1.0
```

I also removed some unnecessary newlines from the error messages and fixed a glitch in the handling of the `message` parameter when the `data` parameter is used.
**Before:**
```
In [1]: import tensorflow as tf 
   ...: tf.assert_equal(1., 0., data=[3.], message="My error message")          

---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-1-ef2b26ad5e73> in <module>
      1 import tensorflow as tf
----> 2 tf.assert_equal(1., 0., data=[3.], message="My error message")
                                    [snip!]
InvalidArgumentError: 3.0
```
**After:**
```
In [1]: import tensorflow as tf 
   ...: tf.assert_equal(1., 0., data=[3.], message="My error message")                  
                      
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-1-ef2b26ad5e73> in <module>
      1 import tensorflow as tf
----> 2 tf.assert_equal(1., 0., data=[3.], message="My error message")
[...stack trace continues...]

InvalidArgumentError: My error message
3.0
```

In the process, I replaced a bunch of near-duplicate code and documentation across the `assert_*` functions with a single function (`_binary_assert()` in `check_ops.py`) and common blocks of documentation (`_binary_assert_doc()` and `_unary_assert_doc()` in `check_ops.py`). `check_ops.py` is now about 125 lines shorter.

I added some new regression tests to cover static assertion checks in graph mode and modified some existing tests to account for the new functionality.

I built a local copy of the documentation for the `tf.debugging` package and reviewed all the resulting Markdown files.

### CODE REVIEW

File: tensorflow/python/ops/check_ops.py [hunk 1]
```diff
@@ -145,6 +145,99 @@ def _decorator(func):
...
+                         (x_sum, x_np[:x_sum],
+                          y_sum, y_np[:y_sum]))
+
+        index_and_values_str = ''
+        if x.shape == y.shape and x.shape.as_list():
+          # If the shapes of x and y are the same (and not scalars),
+          # Get the values that actually differed and their indices.
+          # If shapes are different this information is more confusing
+          # than useful.
+          mask = math_ops.logical_not(eq)
```
Comment (iganichev) [inline tensorflow/python/ops/check_ops.py]: 'eq' should be 'test' here.

File: tensorflow/python/kernel_tests/check_ops_test.py [hunk 2]
```diff
@@ -503,6 +509,12 @@ def test_returns_none_with_eager(self):
       x = check_ops.assert_less(t1, t2)
       assert x is None
 
+  def test_static_check_in_graph_mode(self):
+    with context.graph_mode():
+      with self.assertRaisesRegexp(errors.InvalidArgumentError,
+                                   "Custom error message")
+        check_ops.assert_none_equal(1, 1, message="Custom error message")
```
Comment (iganichev) [inline tensorflow/python/kernel_tests/check_ops_test.py]: assert_none_equal -> assert_less

File: tensorflow/python/kernel_tests/check_ops_test.py [hunk 3]
```diff
@@ -1307,6 +1356,13 @@ def test_empty_tensor_doesnt_raise(self):
       out = array_ops.identity(empty)
     self.evaluate(out)
 
+  def test_static_check_in_graph_mode(self):
+    with context.graph_mode():
+      with self.assertRaisesRegexp(errors.InvalidArgumentError,
+                                   "Custom error message")
+        check_ops.assert_non_positive(-1, message="Custom error message")
```
Comment (iganichev) [inline tensorflow/python/kernel_tests/check_ops_test.py]: -1 is non_positive. If this test passes, assert_non_positive has a bug.

**Review by iganichev — COMMENTED** [review]
Thanks for making check_ops.py much nicer!!!

It looks like most of my comments are addressed in subsequent commits. In the future, it would save some reviewer time if you squash commits where you fix issues of previous commits. Please do this for this chain as well so that we have a clean tree after every commit. You can probably just squash all commits into one.

**Comment by frreiss** [comment]
Hmm, something bad happened when I did a manual rebase. Recreating the PR on a new branch.

**Pull request merged**